sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala

```diff
@@ -88,6 +90,9 @@ case class ShuffledHashJoinExec(
       ignoresDuplicatedKey = ignoreDuplicatedKey)
     buildTime += NANOSECONDS.toMillis(System.nanoTime() - start)
     buildDataSize += relation.estimatedSize
+    hashRelation = relation
+    // cleanup build plan resource eagerly since all rows have been consumed
```

Comment:
is it possible to release the memory by hash join itself instead of asking several downstream operators to do so? e.g. release the hash relation if the stream side iterator has been consumed up.

Reply:
I thought about it. It needs more code change if we want to do this in shj itself.

shj does not have a unified scanner compared with smj, so if we want to release hash relation at the end of output itrator, we need to wrap a new Iterator for shj. Then change all join types with codegen non-codegen code path to use the new Iterator. It will be a code refactor.